artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java

```diff
@@ -587,52 +586,48 @@ void deliver() {
          LargeBodyReader context = message.getLargeBodyReader();
          try {
             context.open();
+            final ByteBuf tmpFrameBuf = PooledByteBufAllocator.DEFAULT.directBuffer(frameSize);
+            final NettyReadable nettyReadable = new NettyReadable(tmpFrameBuf);
             try {
+
                context.position(position);
                long bodySize = context.getSize();
-
-               ByteBuffer buf = ByteBuffer.allocate(frameSize);
+               // materialize it so we can use its internal NIO buffer
+               tmpFrameBuf.ensureWritable(frameSize);
 
                for (; sender.getLocalState() != EndpointState.CLOSED && position < bodySize; ) {
                   if (!connection.flowControl(this::resume)) {
                      context.close();
                      return;
                   }
-                  buf.clear();
-                  int size = 0;
-
-                  try {
-                     if (position == 0) {
-                        replaceInitialHeader(deliveryAnnotationsToEncode, context, WritableBuffer.ByteBufferWrapper.wrap(buf));
-                     }
-                     size = context.readInto(buf);
-
-                     sender.send(new ReadableBuffer.ByteBufferReader(buf));
-                     position += size;
-                  } catch (java.nio.BufferOverflowException overflowException) {
-                     if (position == 0) {
-                        if (log.isDebugEnabled()) {
-                           log.debug("Delivery of message failed with an overFlowException, retrying again with expandable buffer");
-                        }
-                        // on the very first packet, if the initial header was replaced with a much bigger header (re-encoding)
-                        // we could recover the situation with a retry using an expandable buffer.
-                        // this is tested on org.apache.activemq.artemis.tests.integration.amqp.AmqpMessageDivertsTest
-                        size = retryInitialPacketWithExpandableBuffer(deliveryAnnotationsToEncode, context, buf);
+                  tmpFrameBuf.clear();
+                  int headerSize = 0;
+                  int readLimit = frameSize;
+                  if (position == 0) {
+                     replaceInitialHeader(deliveryAnnotationsToEncode, context, new NettyWritable(tmpFrameBuf));
+                     headerSize = tmpFrameBuf.writerIndex();
+                     if (headerSize <= frameSize) {
+                        readLimit = frameSize - headerSize;
```

Comment:
This still feels unnecessarily complicated. It complicates the whole remainder of the method, for a corner case thats actually quite unlikely to even arise. The variable is also confusingly named since it is absolutely not the 'header size', its only non-header stuff than can ever get it into this situation.

Same comment as yesterdays prior version, why not just send the generated first payload at this point (as the old version did when the initial buffer wasnt big enough, in retryInitialPacketWithExpandableBuffer()) and then replace the buffer with an appropriately initial-sized one, and continue the loop (assuming there is more to be sent). That would consolidate all the complexity inside this sub-part of the 'if position=0' handling, and leave the rest of the handling as the most trivial 'read data, send data' as it can be.

Reply:
>  The variable is also confusingly named since it is absolutely not the 'header size', its only non-header stuff than can ever get it into this situation.

I can rework on naming to better capture what's happening here

> why not just send the generated first payload at this point (as the old version did when the initial buffer wasnt big enough, in retryInitialPacketWithExpandableBuffer())

I think it can be done, indeed, meaning that we will skip reading the file at this round, let me rework to do it, I like it